lib/basic_auth.rb

```diff
@@ -0,0 +1,40 @@
+class BasicAuth
+  class << self
+    def authenticate(username, password)
+      return false unless [username, password].all?(&:present?)
+
+      user = credentials.find { |c| c[:username] == username && c[:password] == password }
+      User.new(user[:username], user[:role]) if user.present?
+    end
+
+    def credentials
+      # rubocop:disable Style/ClassVars:
+      @@credentials ||= http_auth.split(",").map do |credential|
+        username, password, role = credential.split("|")
+        { username: username, password: password, role: role }.with_indifferent_access
+      end
+      # rubocop:enable Style/ClassVars:
+    end
+
+    def http_auth
+      Rails.application.config.x.http_auth || ""
+    end
+
+    def env_requires_auth?
+      # Site-wide authentication present in all production-like
+      # environments, but not in production itself.
+      !Rails.env.production? && !Rails.env.test? && !Rails.env.development?
+    end
+  end
+end
+
+class User
```

Comment:
Can we pull this out into its own file? I would be inclined to move some of the `author?` and `publisher?` methods in your controller into the `User` model as well and expose a `@user` for the view instead of having `helper_method`s in your controller

Reply:
Cool, should this live in `app` rather than `lib`?